src/libraries/Common/src/System/Net/Http/Http3/QPack/QPackDecoder.cs

```diff
@@ -0,0 +1,525 @@
+        private const byte LiteralHeaderFieldPrefixMask = 0x0F;
+        private const int LiteralHeaderFieldPrefix = 4;
+
+        //0   1   2   3   4   5   6   7
+        //+---+---+---+---+---+---+---+---+
+        //| 0 | 0 | 0 | 0 | N |NameIdx(3+)|
+        //+---+---+---+---+---+-----------+
+        //| H |     Value Length(7+)     |
+        //+---+---------------------------+
+        //|  Value String(Length bytes)  |
+        //+-------------------------------+
+        private const byte LiteralHeaderFieldPostBasePrefixMask = 0x07;
+        private const int LiteralHeaderFieldPostBasePrefix = 3;
+
+        //0   1   2   3   4   5   6   7
+        //+---+---+---+---+---+---+---+---+
+        //| 0 | 0 | 1 | N | H |NameLen(3+)|
+        //+---+---+---+---+---+-----------+
+        //|  Name String(Length bytes)   |
+        //+---+---------------------------+
+        //| H |     Value Length(7+)     |
+        //+---+---------------------------+
+        //|  Value String(Length bytes)  |
+        //+-------------------------------+
+        private const byte LiteralHeaderFieldWithoutNameReferenceHuffmanMask = 0x08;
+        private const byte LiteralHeaderFieldWithoutNameReferencePrefixMask = 0x07;
+        private const int LiteralHeaderFieldWithoutNameReferencePrefix = 3;
+
+        private const int StringLengthPrefix = 7;
+        private const byte HuffmanMask = 0x80;
+
+        private const int DefaultStringBufferSize = 64;
+
+        private readonly int _maxHeadersLength;
+        private readonly DynamicTable _dynamicTable;
+        private State _state = State.Ready;
+
+        private byte[] _stringOctets;
+        private byte[] _headerNameOctets;
+        private byte[] _headerValueOctets;
+
+        // TODO idk what these are for.
+        private int _requiredInsertCount;
+        //private int _insertCount;
+        private int _base;
+
+        // s is used for whatever s is in each field. This has multiple definition
+        private bool _s;
+        private bool _huffman;
+        private int? _index;
+
+        private byte[] _headerName;
+        private int _headerNameLength;
+        private int _headerValueLength;
+        private int _stringLength;
+        private int _stringIndex;
+        private readonly IntegerDecoder _integerDecoder = new IntegerDecoder();
+
+        private static ArrayPool<byte> Pool => ArrayPool<byte>.Shared;
+
+        // Decoders are on the http3stream now, each time we see a header block
+        public QPackDecoder(int maxDynamicTableSize, int maxHeadersLength)
+            : this(maxDynamicTableSize, maxHeadersLength, new DynamicTable(maxDynamicTableSize)) { }
+
+
+        // For testing.
+        internal QPackDecoder(int maxDynamicTableSize, int maxHeadersLength, DynamicTable dynamicTable)
+        {
+            _maxHeadersLength = maxHeadersLength;
+            _dynamicTable = dynamicTable;
+
+            _stringOctets = Pool.Rent(DefaultStringBufferSize);
```

Comment:
Nit: consider making these lazy? (I see you changed the size to be 64, which is a lot better 😄 )

Reply:
I think this may be useful for responses consisting solely of static table entries, but don't know how frequent that is. Does Kestrel use any non-static headers (Server etc.)? Left as TODO.